torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -385,12 +384,86 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
     // struct are empty, and there is no pre-existing accumulation tensor.
     // Directly assign the sparse tensor to the `contents` field.
     replica.contents = grad;
-    replica.contents.div_(process_group_->getSize());
+    replica.contents.div_(divFactor);
     // The grad is modified in place and needs to be written back.
     return true;
   });
 }
 
+size_t Reducer::getNumBuckets() const {
+  return buckets_.size();
+}
+
+std::vector<at::Tensor> Reducer::getTensorsForBucket(int i) const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  TORCH_CHECK(
+      i < buckets_.size(),
+      c10::str(
+          "Attempt to access bucket index ",
+          i,
+          " but only have ",
+          buckets_.size(),
+          " buckets."))
+  auto& bucket = buckets_[i];
+  std::vector<at::Tensor> tensors;
+  tensors.reserve(bucket.replicas.size());
+  for (const auto& rep : bucket.replicas) {
+    tensors.push_back(rep.contents);
+  }
+  return tensors;
+}
+
+bool Reducer::shouldRebuildBuckets() const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  // if find_unused_parameters_, we do not rebuild buckets.
+  return !find_unused_parameters_ && !has_rebuilt_bucket_;
+}
+
+void Reducer::setForwardPassWorkHandle(
+    std::shared_ptr<c10d::ProcessGroup::Work> forwardPassWorkHandle,
+    at::Tensor& tensor) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  // If there was a previous handle, it should have already completed. This is
+  // because we await this handle in the backward pass, and only reinstall in
+  // the next forward pass.
+  if (forwardPassWorkHandle_.workHandle_) {
+    TORCH_INTERNAL_ASSERT(forwardPassWorkHandle_.workHandle_->isCompleted());
+  }
```

Comment:
It might not have completed if there was an exception during the last forward pass?

Reply:
In this scenario, is the assumption that the user caught the exception in the forward pass and continued training? If this is the case, then I think we'd see the issue in the `backward` corresponding to the last forward pass, since we would need to wait on this work to compute the division factor in the backward pass. So if we successfully run backward for the last fwd pass, then this should always be true. 

Also, if the exception happened in the model's forward that the user defined, then this forward pass allreduce should still have been kicked off (we do this allreduce and forward param sync before calling the user's forward function). In that case, I'm not sure if I see why there is a risk of it not completing?

Reply:
The scenario I had in mind here was that the user call's a forward pass that kicks off this allreduce and we store the handle, after that there is an exception thrown in the forward pass. This exception is caught by the user and they user throws away that batch and calls the forward pass again on a new batch. In this case, it's possible the allreduce hasn't completed yet and this assert would fail and would complain about a bug in PyTorch.

I'm wondering whats the harm in always just overwriting the workHandle_ without having this check?

Reply:
I see, so in this case the batch is thrown out and we immediately go to the next batch, without calling backward() in between. In this case, I think this issue could happen. Moreover, if the work handle wasn't completed in the corresponding backwards pass, then we would throw an error/hang there. So I think we can remove this.